imap/jmap_mail.c

```diff
@@ -9400,15 +9417,17 @@ static int _email_flagupdate_cb(const conv_guidrec_t *rec, void *rock)
     /* Fetch record */
     struct mailbox *mbox = NULL;
     msgrecord_t *mrw = NULL;
-    uint32_t system_flags = 0;
+    uint32_t system_flags = 0, expunged = 0;
 
     int r = jmap_openmbox(req, rec->mboxname, &mbox, /*write*/1);
     if (r) return r;
     r = msgrecord_find(mbox, rec->uid, &mrw);
     if (r) goto done;
     r = msgrecord_get_systemflags(mrw, &system_flags);
     if (r) goto done;
-    if (system_flags & (FLAG_EXPUNGED|FLAG_DELETED)) goto done;
+    r = msgrecord_get_expungedflag(mrw, &expunged);
+    if (r) goto done;
+    if ((system_flags & FLAG_DELETED) || expunged) goto done;
 
     /* Determine if to patch or reset flags */
```

Comment:
I wonder if there's any unwanted side-effects here? Before, the JMAP code copied all system flags, including the internal ones. Now, it looks as if it's just copying the non-internal flags. Which _could_ be the right thing to do. But I'm not entirely sure. Did the JMAP Mail tests all pass in Cassandane with this PR?

Reply:
There are a few failures that seem unrelated, so I'm working on them. I wanted your review on the patch while I figure out the issues with Cassandane.